### PR TITLE
feat(CX-3348): Update Input components

### PR DESCRIPTION
**Pull request by MrSltun**
This PR resolves [CX-3348]

> [!IMPORTANT]
> This is a major version update, please take a look at **Features & Breaking Changes** section

## Description
This PR updates the input components in Palette to match the new design

## Features & Breaking Changes
- Input, LabelledInput, Select, MultiSelect, and TextArea now come with a title that floats when active/completed 
- New PhoneInput element that includes a searchable list of country codes and an input
- Input now has a new prop called `showCounter` to render the character limit under the Input

### ⚠️ Breaking Changes ⚠️
- Select element prop `variant` is no longer supported
- The initial Select value should be an empty string (`""`) to enable title animation on select
- Adding `title` to an element will add 10px top padding to the element, preventing the title from being cut

### New Input ✨
<details>
<summary>Details</summary>

#### Screenshots
| Status | Screenshot |
| -- | -- |
| default | image |
| with title | image |
| with description | image image |
| with a placeholder (no title) | image |
| with a placeholder (with title - focused) | image |
| required | image |
| hover | image |
| focused | image |
| finished typing | image |
| disabled | image |
| with error message | image |
| with character counter | image image |
| with prefix and title | image |
| with prefix and title - focused | image |
| with suffix and title | image | 
| with suffix and title - focused | image |

#### Demos of Moving Title
| Input | Video |
| -- | -- |
| Normal Input | <video src="https://user-images.githubusercontent.com/20655703/216614557-f475e0e9-bd55-406e-9e72-27f4cce7edd5.mov" /> |
| Labelled input | <video src="https://user-images.githubusercontent.com/20655703/216615036-709cc9c2-1c0c-4fc8-bb77-5400d355b423.mov" /> |
</details>

### New Phone Input ✨

<details><summary>Details</summary>

#### Screenshots
| Status | Screenshot |
| --- | --- |
| Empty | image |
| With a placeholder | image |
| Disabled | image |
| Filled | image |
| Hover | image |
| Focused | image |
| Required | image |
| With the list open | image |
| With the list open and filtered | image |
| With errors | image |

#### Demo

| Jounrey | Video |
| -- | -- |
| ⌨️ & 🖱️ | <video src="https://user-images.githubusercontent.com/20655703/233412661-97e58cc1-2632-4229-ac42-f739c2cd3774.mov" width="670" /> | 
| ⌨️ ONLY | <video src="https://user-images.githubusercontent.com/20655703/233412825-04ad00c7-db69-4047-a50f-12a24c0fa5c4.mov" /> |

</details> 

### New TextArea ✨
<details><summary>Details</summary>

#### Screenshots
| Status | Screenshot |
| -- | -- |
| default | image |
| with title | image |
| with description | image image |
| with placeholder (no title) | image |
| with placeholder (with title - focused) | image |
| required | image |
| hover | image |
| focused | image |
| finished typing | image |
| disabled | image |
| disabled (with value) | image |
| with error message | image |
| with character counter | image image |
| with character counter (over limit) | image |

</details> 

### New Select ✨

<details><summary>Details</summary>

#### Screenshots
| Status | Screenshot |
| -- | -- |
| default | image |
| default (with value) | image |
| with title | image |
| with title (with value) | image |
| with title and description | image image |
| required | image |
| hover | image |
| focused | image |
| with error message | image |
| disabled | image |
| disabled (with value) | image |

#### Demo Video
<video src="https://github.com/artsy/palette/assets/20655703/b6ae8eed-8371-437d-9ad9-7303dfef3378" />

</details> 

### New MultiSelect ✨

<details><summary>Details</summary>

#### Screenshots
| Status | Screenshot |
| -- | -- |
| default | image |
| default (with value) | image |
| with title | image |
| with title (with value) | image |
| with title and description | image image |
| required | image |
| hover | image |
| focused | image |
| with error message | image |
| disabled | image |

#### Demo Video
<video src="https://github.com/artsy/palette/assets/20655703/fa1d8df4-a2fb-4454-b54b-ce41915a99c6" />	

</details> 

[CX-3330]: https://artsyproduct.atlassian.net/browse/CX-3330?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @artsy/palette-charts@36.0.0-canary.1250.29384.0
  npm install @artsy/palette@37.0.0-canary.1250.29384.0
  # or 
  yarn add @artsy/palette-charts@36.0.0-canary.1250.29384.0
  yarn add @artsy/palette@37.0.0-canary.1250.29384.0
  ```
</details>

[CX-3348]: https://artsyproduct.atlassian.net/browse/CX-3348?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by MrSltun** [comment]
And as a side-effect, LabeledInput changed to match the new design as well 😄

image

Let's hope I didn't break anything else 🌚

**Review by MounirDhahri — APPROVED** [review]
niiice! thanks Sultan.
@dzucconi Should we wait for the other input fields (credit card and money) to be done before releasing this to users to avoid UI conflicts ?

**Comment by MrSltun** [comment]
@MounirDhahri I'm merging this to a working branch called [`in-progress/new-input`](https://github.com/artsy/palette/tree/in-progress/new-input) as @dzucconi suggested, so that we can proceed with the work and release a canary version of Palette from the working branch to test it on Force before releasing it to the public :)

Please correct me if I got it wrong! @dzucconi

**Comment by MounirDhahri** [comment]
Sounds great, forgot that we spoke about it Already! let's proceed with that then once Damon approves

**Comment by dzucconi** [comment]
Yeah: we can't merge this until _every_ input-like component is converted and we validate Force. (I've added the `canary` tag).

(Looking now though!)

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:18]: I don't think we should support a configurable variant on this. They should all be the same. Safe to remove this prop.

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:59]: `style` No need for the braces

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:63]: 👍 The `Tooltip` component clones this child and adds `tabIndex={0}` to ensure this is keyboard focusable.

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:59]: yeah, that one slipped 😅😬

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:50]: `suggestion` It's possible we want to make this a required prop rather than a fallback. I don't really think there should be a fallback value though, either way.

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:122]: Is there a spec for the animation? As far as transition values go, etc.

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:175]: Same question regarding spec. Where is this curve coming from?

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:175]: We don't want to `transition` *all* properties; so make sure you add `transform` here.

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:171]: Hm, is there a way to get at our token value — I know `0.5` has some issues. `themeGet("space['0.5']")` maybe?

File: packages/palette/src/elements/Label/Label.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Label/Label.tsx:1]: 🙏

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:139]: Smart!

**Review by dzucconi — COMMENTED** [review]
Looking really good!

One minor general comment I have is that we might want to `opacity` transition the placeholder in, simultaneously as the label transitions its `translation`, so that there isn't an overlap.

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:50]: Actually, t's totally valid, of course, to have inputs without names, for instance when they are part of a non-form component — like an autocomplete UI or something.

Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:50]: I'm concerned if we have multiple inputs on a page without names, would the labels be associated with the same input?

example:

```html
<input name="palette-input" ... />
<label for="palette-input" ...>

...

<input name="palette-input" ... />
<label for="palette-input" ...>

```

Is that even possible to have two inputs with the same name? and if so, should we add a random digit at the end of the fallback input name?

**Review by dariakoko — APPROVED** [review]
Great work Sultan!

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:122]: I don't think there are any specific specs for the animation speed, so I'll change it to be .25 to match the current speed we have 👍

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:175]: For the curve, I was playing around with [this tool](https://cubic-bezier.com) to be as close to [the material design guidelines](https://m3.material.io/styles/motion/easing-and-duration/applying-easing-and-duration#d4348fa7-87b9-4dda-be70-c126111fd63c) 

But for the transition, I'll specify the properties that are going to be changed :) (font size, color, left, top, etc)

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:175]: Wondering if we can do this label position transition using *only* `transform`, actually? There's no real reason we couldn't. See: https://web.dev/animations-guide/#layout

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:171]: I tried multiple ways, but none were successful 😞

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:171]: We should probably just introduce some named aliases for these, like `space.half-x` or something. Anyway; fine for now.

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:171]: (You also could just pass these using `attrs`)

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:175]: I managed to update the size (using `scale`) and the position (using `translate(x , y)`), but I'm not sure how to animate the changing of the text color with transform, do you know how we could do that?

In the meantime, I'll update the transition property to only update the animating ones, not _all_

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:175]: Nice — you *could* get clever using `opacity` since that is also one of the properties that's hardware accelerated but for now I'd continue to just transition on `color`. We might want to tag these with [`will-change`](https://developer.mozilla.org/en-US/docs/Web/CSS/will-change), but it'd probably be good to just measure and optimize later.

I think these animations are fast enough and nothing else generally will be happening on screen as far as layout goes so I'm not anticipating this being much of a problem. Just, in general, if something can be achieved with a `transform` then we should do that. 👍

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [inline packages/palette/src/elements/Input/Input.tsx:215]: Changed these to fix this issue from showing up

### Before
image

After: 
image

**Comment by dzucconi** [comment]
Was wondering what happened to this — can you rebase though instead of the merge?

File: packages/palette/src/elements/Input/tokens.ts
Comment (damassi) [inline packages/palette/src/elements/Input/tokens.ts:89]: All of the animation styling in here look great (checked out [the review app](https://new-input.artsy.net/)) 👍 

I wonder if there's any way we might be able to improve the perf a bit tho. It _could_ just be my machine, but it feels like the frame rate is a bit slow (or rather, could be faster). I know translate Y pushes to the gpu, but 🤔 

@MrSltun , @dzucconi - thoughts? Maybe its just fine... not too sure. 


https://github.com/artsy/palette/assets/236943/2273bffb-d8dc-4d81-9915-dba848da7e26

Comment (dzucconi) [reply packages/palette/src/elements/Input/tokens.ts:89]: I'd be interested to know if the color is what's slowing you down; otherwise no.

Totally fine on my M2 with 32GB of ram though 😆

Comment (damassi) [reply packages/palette/src/elements/Input/tokens.ts:89]: Oh interesting, i just went back to the tab and its all ok! It was due to low battery on my computer. Looks perfect now. 

Ignore!

Comment (damassi) [reply packages/palette/src/elements/Input/tokens.ts:89]: So eager for this to launch ⭐

File: packages/palette/src/elements/Input/Input.tsx
Comment (damassi) [inline packages/palette/src/elements/Input/Input.tsx:88]: #superminor - 
```tsx
if (...) {
  // use brackets :) 
}
```

**Review by damassi — APPROVED** [review]
Nice work @MrSltun - and your CSS skills are impressive :)

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:27]: I feel strongly that we need to get rid of these props. We extend BoxProps so just use them — I'm confused by this

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:57]: Again: we really cannot have a default name here.

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [inline packages/palette/src/elements/Input/Input.tsx:171]: I'm going to need some kind of visual explanation for the existence of this?

**Comment by damassi** [comment]
One thing I noticed @MrSltun is that checking out the order flow in the review app is that the drop down doesn't seem to be selectable: 

Screenshot 2023-10-25 at 10 25 42 AM

**Comment by damassi** [comment]
@dzucconi - do you have any blocking comments, or can we do a follow-up here?

**Comment by dzucconi** [comment]
I would really address the prefixOffset props stuff before merging. Changing props APIs down the road is hard.

We also need to drop the default name

**Comment by damassi** [comment]
No prob, just wanted to be sure 👍 I shall contain my eagerness

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:27]: These were added to support `LabeledInput` since we used to have a default offset the label that was only added to the right of the element


In the new designs we have two options for labels, either left or right, and since we use our `Input` for it, we need to add an offset for the input and for the label positioning if the label is on the left

This is where it is used: [code](https://github.com/artsy/palette/blob/mrsltun/CX-3330/new-input-component/packages/palette/src/elements/LabeledInput/LabeledInput.tsx#L36)

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:27]: I'd probably just split off `pl` and `pr` though? We don't want these exposed publicly. We could also dunder name them to discourage usage.

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx:27]: Why accept arbitrary values if we know the exact contexts in which they are used?

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx]: Just removed the default name :)

File: packages/palette/src/elements/Input/Input.tsx
Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx:171]: This will be removed

File: packages/palette/src/elements/Input/Input.tsx
Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx]: Cool! Does the `labelOffset` actually change? Or can it just be a boolean?

Comment (MrSltun) [reply packages/palette/src/elements/Input/Input.tsx]: It's a number that adds a padding-left to the label, and that number depends on the width of the text in the `LabeledInput`

Comment (dzucconi) [reply packages/palette/src/elements/Input/Input.tsx]: Ah OK — got it thanks

**Review by dzucconi — APPROVED** [review]

**Comment by damassi** [comment]
Alright @MrSltun - thanks for all of your hard work on this, for real. Gonna merge it and we can follow up with a merge on the review app tomorrow and I can update Volt 2 first, then Forque and Volt 1.

**Comment by artsyit** [comment]
:rocket: PR was released in [`@artsy/palette-charts@36.0.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette-charts%4036.0.0), [`@artsy/palette@37.0.0`](https://github.com/artsy/palette/releases/tag/%40artsy/palette%4037.0.0) :rocket:

**Pull request merged**
